lib/proxy/server_test.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// newSelfSignedCA creates a new CA for testing.
+func newSelfSignedCA() (*tlsca.CertAuthority, error) {
+	rsaKey, err := ssh.ParseRawPrivateKey(fixtures.PEMBytes["rsa"])
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cert, err := tlsca.GenerateSelfSignedCAWithSigner(
+		rsaKey.(*rsa.PrivateKey), pkix.Name{}, nil, defaults.CATTL,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ca, err := tlsca.FromCertAndSigner(cert, rsaKey.(*rsa.PrivateKey))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return ca, nil
+}
+
+// certFromIdentity creates a tls config for a given CA and identity.
+func certFromIdentity(t *testing.T, ca *tlsca.CertAuthority, ident tlsca.Identity) *tls.Config {
+	if ident.Username == "" {
+		ident.Username = "test-user"
+	}
+
+	subj, err := ident.Subject()
+	require.NoError(t, err)
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	clock := clockwork.NewRealClock()
+
+	request := tlsca.CertificateRequest{
+		Clock:     clock,
+		PublicKey: privateKey.Public(),
+		Subject:   subj,
+		NotAfter:  clock.Now().UTC().Add(time.Minute),
+		DNSNames:  []string{"127.0.0.1"},
+	}
+	certBytes, err := ca.GenerateCertificate(request)
+	require.NoError(t, err)
+
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(privateKey)})
+	cert, err := tls.X509KeyPair(certBytes, keyPEM)
+	require.NoError(t, err)
+
+	pool := x509.NewCertPool()
+	pool.AddCert(ca.Cert)
+
+	config := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	}
+
+	return config
+}
+
+type mockAccessCache struct {
+	auth.AccessCache
+}
+
+// TestServerTLS ensures that only trusted certificates with the proxy role
+// are accepted by the server.
+func TestServerTLS(t *testing.T) {
+	ca1, err := newSelfSignedCA()
+	require.NoError(t, err)
+
+	ca2, err := newSelfSignedCA()
+	require.NoError(t, err)
+
+	tests := []struct {
+		desc      string
+		server    *tls.Config
+		client    *tls.Config
+		expectErr bool
+	}{
+		{
+			desc: "trusted certificates with proxy roles",
+			server: certFromIdentity(t, ca1, tlsca.Identity{
+				Groups: []string{string(types.RoleProxy)},
+			}),
+			client: certFromIdentity(t, ca1, tlsca.Identity{
+				Groups: []string{string(types.RoleProxy)},
+			}),
+			expectErr: false,
+		},
+		{
+			desc: "trusted certificates with incorrect server role",
+			server: certFromIdentity(t, ca1, tlsca.Identity{
+				Groups: []string{string(types.RoleAdmin)},
+			}),
+			client: certFromIdentity(t, ca1, tlsca.Identity{
+				Groups: []string{string(types.RoleProxy)},
+			}),
+			expectErr: true,
+		},
+		{
+			desc: "certificates with correct role from different CAs",
```

Comment:
I think this particular test case does not really represent what happens in the code. 
The expected error is the result of the way the test is setup rather than actually be a legitimate error. I've removed it from my client branch as I was reworking the tests a bit after I implemented the equivalent of getConfigForClient on the client side.

Reply:
https://github.com/gravitational/teleport/pull/10440/files#diff-8668137b10a4a8bf67506307e688582892039b3bd03ac2090b20ad0da9b9a3d1R49-R56

Reply:
Not sure I understand the issue. This is specifically testing the server side cert validation. Could we leave this and have a separate test for the client side?

One change I might add is setting the `tc.client.InsecureSkipVerify = true` to ensure the failure happens server side.

Reply:
Unless I am seeing this wrong, there's nothing in the code itself that prevents a connection between a client and a server with certs from two different CAs. It is just triggered by the way the test case is implemented.

Reply:
The underlying call to ClientHandshake/ServerHandshake should return an error in this case.